Conserve bandwidth when using ManagedFile

a3f opened the pull request:
This PR aims at improving user experience with ManagedFile:

* use `rsync --compress=1 --sparse` to speed up sparse file upload
* heuristic for determining file is on NFS and foregoing the preceding transfer if it is

**What I intend to use this feature for:**

Save time when using labgrid-client.

**How does labgrid benefit as a testing library from the feature?**

The added `run`/`run_check` command can be used in tests where running commands on the exporter system is required. 

**How did you verify the feature works?**

I included some tests.

**Description**

<!---
This checklist roughly outlines the steps for new features, remove and add tasks as needed:
--->
**Checklist**
- [x] Documentation for the feature
- [x] Tests for the feature 

<!---
A library feature which other developers can use:
--->
- [x] Add a section on how to use the feature to doc/development.rst
<!---
Provide a short summary for the CHANGES.rst file
--->
- [ ] CHANGES.rst has been updated
<!---
Did you test the change locally? If yes, best to mention how you did it in the description section.
--->
- [x] PR has been tested
<!---
If your PR touched the man pages they have to be regenerated by calling make in the man subdirectory of the project
--->

<!---
In case your PR fixes a bug, please reference it in the next line, i.e.
Fixes #[insert number without brackets here]
--->

jluebbe commented on labgrid/util/ssh.py:
```diff
@@ -191,23 +192,75 @@ def wrapper(self, *_args, **_kwargs):
+            codec (string, optional): output encoding. Defaults to "utf-8".
+            decodeerrors (string, optional): behavior on decode errors. Defaults
+                 to "strict". Refer to stdtypes' bytes.decode for details.
+            timeout (float, optional): currently unused
```
Please don't add unused arguments.

a3f replied:
Removed. I had included them for compatibility with SSHDriver.

jluebbe commented on labgrid/util/ssh.py:
```diff
@@ -124,6 +124,7 @@ class SSHConnection:
         validator=attr.validators.instance_of(str)
     )
     _forwards = attr.ib(init=False, default=attr.Factory(dict))
+    stderr_merge = attr.ib(default=False, validator=attr.validators.instance_of(bool))
```
It seems this should be an option for the actual run command, as we always create separate ssh sessions (in the same shared connection).

a3f replied:
Changed.

jluebbe commented on labgrid/util/managedfile.py:
```diff
@@ -53,12 +48,41 @@ def sync_to_resource(self):
+        localout = local.stdout.decode("utf-8", "backslashreplace").split('\n')
+        localout.pop() # remove trailing empty element
+
+        self._on_nfs_cached = remote[2] == local.returncode == 0 and remote[0] == localout
```
Please use a debug logger for separate error cases.

a3f replied:
A different debug message for each boolean check, you mean? I've done so now.

jluebbe commented on doc/development.rst:
```diff
@@ -405,6 +405,15 @@ To use it in conjunction with a `Resource` and a file:
    path = mf.get_remote_path()
 
+Unless constructed with `ManagedFile(..., detect_nfs=False)`, ManagedFile
+employs following heuristic to check if a file is on NFS and if so, foregoes
```
"employs *the* following"

a3f replied:
done.

jluebbe commented on doc/development.rst:
```diff
@@ -405,6 +405,15 @@ To use it in conjunction with a `Resource` and a file:
 
+Unless constructed with `ManagedFile(..., detect_nfs=False)`, ManagedFile
+employs following heuristic to check if a file is on NFS and if so, foregoes
+the transfer and `get_remote_path()` just returns the local path:
```
perhaps add "(which is identical to the remote path in that case)"

a3f replied:
done.

Bastian-Krause commented on tests/test_util.py:
```diff
@@ -257,8 +294,9 @@ def test_local_managedfile(target, tmpdir):
     )
     hash = hashlib.sha256(t.read().encode("utf-8")).hexdigest()
-    mf = ManagedFile(t, res)
+    mf = ManagedFile(t, res, detect_nfs = False)
```
No space allowed around keyword argument assignment.

a3f replied:
Removed.

Bastian-Krause commented on tests/test_util.py:
```diff
@@ -84,7 +91,23 @@ def test_sshconnection_connect(connection_localhost):
+
+@pytest.mark.localsshmanager
+def test_sshconnection_run_merged_stderr(connection_localhost_merged_stderr):
+    (stdout, stderr, exitcode) = connection_localhost_merged_stderr.run("echo stderr >&2; echo stdout")
```
The brackets before `=` are obsolete.

a3f replied:
removed.

Bastian-Krause commented on tests/test_util.py:
```diff
@@ -84,7 +91,23 @@ def test_sshconnection_connect(connection_localhost):
+
+@pytest.mark.localsshmanager
+def test_sshconnection_run_fail(connection_localhost):
+    (stdout, stderr, exitcode) = connection_localhost.run("false")
```
Here the same.

a3f replied:
removed as well.

a3f commented on labgrid/util/managedfile.py:
```diff
@@ -53,12 +48,41 @@ def sync_to_resource(self):
             conn = sshmanager.open(host)
-            conn.run_command("mkdir -p {}".format(self.rpath))
+
+            if self._on_nfs(conn) == True:
```
Done.

a3f commented on labgrid/util/managedfile.py:
```diff
@@ -53,12 +48,41 @@ def sync_to_resource(self):
+        if not self.detect_nfs:
+            return False
+
+        fmt="inode=%i,size=%s,birth=%W,modified=%Y"
```
Done.

a3f commented on labgrid/util/ssh.py:
```diff
@@ -191,23 +192,76 @@ def wrapper(self, *_args, **_kwargs):
     @_check_connected
-    def run_command(self, command):
+    def run(self, command, *, codec = "utf-8", decodeerrors = "strict",
+            timeout = None): # pylint: disable=unused-argument
```
Fixed.

a3f commented on tests/test_util.py:
```diff
@@ -257,8 +294,9 @@ def test_local_managedfile(target, tmpdir):
 
     assert hash == mf.get_hash()
     assert str(t) == mf.get_remote_path()
+
```
Oh, removed now!

a3f commented:
I've fixed the mentioned issues. I've since noticed that run/run_check aren't a drop in replacement for run_command, as run_command doesn't redirect the stdout/stderr of its child process and thus the output is displayed on the connected terminal.

To support this, I've added three more parameters to `run`:

            force_tty (bool, optional): force allocate a tty (ssh -t). Defaults
                 to False
            stdout_loglevel (int, optional): log stdout with specific log level
                 as well. Defaults to None, i.e. don't log.
            stderr_loglevel (int, optional): log stderr with specific log level
                 as well. Defaults to None, i.e. don't log.

ssh `-tt` is necessary so the remote subprocesses don't fully buffer their input and 'real-time' logging 
 to `stdout_loglevel` and `stderr_loglevel` is possible.

Bastian-Krause commented on labgrid/util/ssh.py:
```diff
@@ -191,23 +192,110 @@ def wrapper(self, *_args, **_kwargs):
+        stdout, stderr, exitcode = self.run(command, codec=codec,
+                decodeerrors=decodeerrors, force_tty=force_tty,
+                stderr_merge=stderr_merge, stdout_loglevel=stdout_loglevel,
+                stderr_loglevel=stderr_loglevel)
```
Wrong continued indentation.

a3f replied:
How should it look like?

Bastian-Krause commented on labgrid/util/ssh.py:
```diff
@@ -191,23 +192,110 @@ def wrapper(self, *_args, **_kwargs):
+
+    def run_check(self, command, *, codec="utf-8", decodeerrors="strict",
+        force_tty=False, stderr_merge=False, stderr_loglevel=None,
+        stdout_loglevel=None):
```
Wrong continued indentation.

Bastian-Krause commented on labgrid/util/ssh.py:
```diff
@@ -191,23 +192,110 @@ def wrapper(self, *_args, **_kwargs):
+                    if loglevel is not None:
+                        self._logger.log(loglevel, line)
+
+        return stdout, stderr, sub.wait()
```
Is all of this really necessary? Seems much more compilcated than the SSHDriver.

a3f replied:
I want the function to be synchronous, (like `subprocess.check_output`/0, but display output in "real-time" (like if stdout/stderr is going to a terminal). If I do it like in the SSHDriver, I won't get output until the process terminates, which is annoying if the process is long-running. 

For that reason, I redirect stdout/stderr into pipes, but now the libc will do block-buffering, thus I provide a way to specify `-tt`, which forces allocation of a PTY, so the output streams are line buffered. As the buffer for stdout/stderr is limited, it must be cleared regularly. A simple `stdout.readline()` followed by a `stderr.readline()` and bail if both are at EOF won't cut it, because it can deadlock: subprocess fills the stderr buffer and blocks while parent waits to read a line from the stdout buffer. For that reason, the `select` was necessary.
`subprocess` uses threads instead behind the scene to achieve this. I guess because it allows a unified implementation for *nix and Windows (Windows' select does only work on sockets and not file descriptors in general).

It might make sense for SSHDriver to do the same as well..
